source/common/buffer/watermark_buffer.h

```diff
@@ -0,0 +1,64 @@
+#pragma once
+
+#include <string>
+
+#include "envoy/buffer/buffer.h"
+
+namespace Envoy {
+namespace Buffer {
+
+// A wrapper for an underlying buffer which does watermark validation.
+// The underlying buffer's ownership is transfered to the Watermark buffer.   Each time the inner
+// buffer is resized (written to or drained), the watermarks are checked.  As the buffer size
+// transitions from under the low watermark to above the high watermark, the above_high_watermark
+// function is called one time. It will not be called again until the buffer is drained below the
+// low watermark, at which point the below_low_watermark function is called.
+class WatermarkBuffer : public Instance {
```

Comment:
This is really nice @alyssawilk. The only question I had is whether it makes sense to make this a subclass of `OwnedImpl` or not? It's probably better as a wrapper, since it has nothing really to do with the `OwnedImpl` implementation, OTOH you can avoid a few lines of wrapper calls where it would just chain to the parent class. I think what you have is good, if you can add the tests to provide coverage I think we're good to go.